jooq/src/main/java/io/micronaut/configuration/jooq/JooqConfigurationFactory.java

```diff
@@ -0,0 +1,53 @@
+package io.micronaut.configuration.jooq;
+
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.annotation.EachBean;
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Parameter;
+import io.micronaut.inject.qualifiers.Qualifiers;
+import org.jooq.*;
+import org.jooq.conf.Settings;
+import org.jooq.impl.DefaultConfiguration;
+import org.jooq.impl.DefaultDSLContext;
+
+import javax.sql.DataSource;
+
+@Factory
+public class JooqConfigurationFactory {
+
+    @EachBean(DataSource.class)
+    public DefaultConfiguration jooqConfiguration(
+            @Parameter String name,
+            DataSource dataSource,
+            ApplicationContext ctx
+    ) {
+        DefaultConfiguration configuration = new DefaultConfiguration();
+
+        JooqConfigurationProperties properties = ctx.findBean(JooqConfigurationProperties.class, Qualifiers.byName(name))
+                .orElseGet(JooqConfigurationProperties::new);
+        configuration.setSQLDialect(properties.determineSqlDialect(dataSource));
+
+        configuration.setDataSource(dataSource);
+        ctx.findBean(TransactionProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setTransactionProvider);
+        ctx.findBean(Settings.class, Qualifiers.byName(name)).ifPresent(configuration::setSettings);
+        ctx.findBean(ExecutorProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setExecutorProvider);
+        ctx.findBean(RecordMapperProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setRecordMapperProvider);
+        ctx.findBean(RecordUnmapperProvider.class, Qualifiers.byName(name)).ifPresent(configuration::setRecordUnmapperProvider);
+        configuration.setExecuteListenerProvider(ctx.getBeansOfType(ExecuteListenerProvider.class, Qualifiers.byName(name))
+                .toArray(new ExecuteListenerProvider[0]));
+        configuration.setRecordListenerProvider(ctx.getBeansOfType(RecordListenerProvider.class, Qualifiers.byName(name))
+                .toArray(new RecordListenerProvider[0]));
+        configuration.setVisitListenerProvider(ctx.getBeansOfType(VisitListenerProvider.class, Qualifiers.byName(name))
+                .toArray(new VisitListenerProvider[0]));
+        configuration.setTransactionListenerProvider(ctx.getBeansOfType(TransactionListenerProvider.class, Qualifiers.byName(name))
+                .toArray(new TransactionListenerProvider[0]));
+
+        return configuration;
+    }
+
+    @EachBean(Configuration.class)
```

Comment:
You may want add `@Context` so it is initialized on startup so the application fails early for errors. We do this for most database related beans

Reply:
Besides `SQLDialect` detection it does not initiate any connections and does not do any configuration validation. On the other hand `@Context` will transitively force all injected provider beans to be eagerly initialized which can slower application startup quite a bit, right?

Reply:
It may do, you can always test the impact on startup time by adding `@Context` and seeing what happens. If it is not required to validate the state of the application then you can leave `@Context` off.